language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -82,79 +111,150 @@ mergePackageMap ps = foldM merge Map.empty ps
+                 putStrLn $ "Generating " <> id <> " as " <> asName
+                 deps <- daml2ts opts pm pkgId pkg mbPkgName damlTypesVersion packageVersion
+                 pure (name, name, deps)
+        whenJust optInputPackageJson $
+          writeTopLevelPackageJson optOutputDir dependencies
 
-dup :: a -> (a, a)
-dup x = (x, x)
+packageNameStr :: PackageId -> Maybe PackageName -> String
+packageNameStr pkgId mbPkgIdent =
+  T.unpack (maybe (unPackageId pkgId) unPackageName mbPkgIdent)
 
-infixr 6 <.> -- This is the same fixity as '<>'.
-(<.>) :: T.Text -> T.Text -> T.Text
-(<.>) u v = u <> "." <> v
+newtype Scope = Scope {unscope :: String}
+newtype Dependency = Dependency {undependency :: String}  deriving (Eq, Ord)
 
-genModule :: Map.Map PackageId (Maybe PackageName, Package) -> PackageId -> Module -> Maybe T.Text
-genModule pm curPkgId mod
-  | null serDefs = Nothing
+-- Gives the scope '@foo' given a directory path like '/path/to/foo'.
+scopeOfScopeDir :: FilePath -> Scope
+scopeOfScopeDir = Scope . ("@" <>) . takeFileName
+
+-- From the path to a package like '/path/to/daml2ts/d14e08'
+-- calculates '@daml2ts/d14e08' suitable for use as the "name" field
+-- of a 'package.json'.
+packageNameOfPackageDir :: FilePath -> String
+packageNameOfPackageDir packageDir = scope <> "/" <> package
+  where
+    scope = unscope $ scopeOfScopeDir (takeDirectory packageDir)
+    package = takeFileName packageDir
+
+-- Write the files for a single package.
+daml2ts :: Options -> Map.Map PackageId (Maybe PackageName, Package) ->
+    PackageId -> Package -> Maybe PackageName -> String -> String -> IO [Dependency]
+daml2ts Options{..} pm pkgId pkg mbPkgName damlTypesVersion packageVersion = do
+    let scopeDir = optOutputDir
+          -- The directory into which we generate packages e.g. '/path/to/daml2ts'.
+        packageDir = scopeDir </> packageNameStr pkgId mbPkgName
+          -- The directory into which we write this package e.g. '/path/to/daml2ts/davl-0.0.4'.
+        packageSrcDir = packageDir </> "src"
+          -- Where the source files of this package are written e.g. '/path/to/daml2ts/davl-0.0.4/src'.
+        scope = scopeOfScopeDir scopeDir
+          -- The scope e.g. '@daml2ts'.
+          -- We use this, for example, when generating import declarations e.g.
+          --   'import * as pkgd14e08_DA_Internal_Template from @daml2ts/d14e08/lib/DA/Internal/Template';'
+    createDirectoryIfMissing True packageSrcDir
+    -- Write .ts files for the package and harvest references to
+    -- foreign packages as we do.
+    T.writeFileUtf8 (packageSrcDir </> "packageId.ts") $ T.unlines
+        ["export default '" <> unPackageId pkgId <> "';"]
+    dependencies <- nubOrd . concat <$> mapM (writeModuleTs packageSrcDir scope) (NM.toList (packageModules pkg))
+    -- Now write 'package.json', 'tsconfig.json' and '.eslint.rc.json'
+    -- files into the package dir. The 'package.json' needs the
+    -- dependencies.
```

Comment:
I'd suggest to write the `packageId.ts` file here, together with the other "meta data".

Reply:
Ok.

Reply:
https://github.com/digital-asset/daml/pull/4858